File: cpp/examples/live_send_header.cpp
// 发送头三个字段（defaultTopic / defaultTopicQueueNums / brokerName）真机验证。
// 用法：rmq_live_send_header 127.0.0.1:9876
//
// 与 Python 的 verify_send_header_live.py、Rust 的 live_send_header.rs、.NET 的
// LiveSendHeader.cs 对齐（H0–H5 一一对应）。
//
// 前置：NameServer + Broker 已起，``autoCreateTopicEnable=true``。
//
// 离线单测（tests/test_send_retry.cpp）锁的是上线形状：V2 头的单字母键 c/d/n 有没有真
// 写进 extFields、五种发送入口是不是带同一份值。但「字段上了线」和「broker 真拿它做了
// 决定」是两件事，后者只有真集群能证：
//   H1  默认值：不带任何配置发到新 topic，broker 按 min(d=4, TBW102.writeQueueNums) 建
//       队列（TopicConfigManager.java:289），与 Java 客户端默认行为一致。
//   H2  setDefaultTopicQueueNums(2) 真的生效：建出来的 topic 只有 2 条队列。修之前这里
//       写死 4，这条必然变 4 —— 那是那个假 setter 唯一可观测的后果。
//   H3  setCreateTopicKey(src) 真的生效：先建一个带 PERM_INHERIT、3 条队列的模板 topic，
//       再以它为 c 发送 → 新 topic 继承模板的 3 条队列，而不是 TBW102 的 8 条。
//   H4  补上这三个字段之后，五种入口（同步 / 定点 / 单向 / 批量 320 / 异步）在真 broker
//       上仍逐条落地，条数一条不差。
//   H5  n（brokerName）：落点就是路由选中的那台 broker 名。⚠ 经典 broker 的发送链路里
//       没有 requestHeader.getBrokerName() 的读者（5.5.1 源码 grep 过），所以 n 在线上
//       的存在只能由离线抓帧证明，这里不假装能观测到它。
#include <algorithm>
#include <atomic>
#include <chrono>
#include <condition_variable>
#include <cstdio>
#include <functional>
#include <map>
#include <memory>
#include <mutex>
#include <string>
#include <thread>
#include <vector>

#include "rocketmq/client/admin.h"
#include "rocketmq/client/exception.h"
#include "rocketmq/client/producer.h"
#include "rocketmq/client/result.h"
#include "rocketmq/common/message.h"
#include "rocketmq/common/mix_all.h"
#include "rocketmq/common/sysflag.h"
#include "rocketmq/remoting/protocol/route.h"

using namespace rocketmq;

namespace {

int32_t gPass = 0;
int32_t gFail = 0;

void check(const std::string& name, bool ok, const std::string& detail = std::string()) {
    if (ok) {
        ++gPass;
    } else {
        ++gFail;
    }
    std::printf("  [%s] %s%s\n", ok ? "PASS" : "FAIL", name.c_str(),
                detail.empty() ? "" : ("  " + detail).c_str());
}

std::string num(int32_t v) { return std::to_string(v); }

std::string num64(int64_t v) { return std::to_string(v); }

int64_t nowMs() {
    return std::chrono::duration_cast<std::chrono::milliseconds>(
               std::chrono::steady_clock::now().time_since_epoch())
        .count();
}

bool waitUntil(const std::function<bool()>& pred, int32_t timeoutMs) {
    const int64_t deadline = nowMs() + timeoutMs;
    while (nowMs() < deadline) {
        if (pred()) return true;
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }
    return pred();
}

Bytes bytesOf(const std::string& s) { return Bytes(s.begin(), s.end()); }

// 一次异步发送的终态
class Latch : public SendCallback {
public:
    void onSuccess(const SendResult& r) override {
        std::lock_guard<std::mutex> lk(m_);
        result_ = r;
        ++oks_;
        cv_.notify_all();
    }
    void onException(const std::exception_ptr& e) override {
        std::lock_guard<std::mutex> lk(m_);
        errors_.push_back(exceptionMessage(e) + " [" + exceptionTypeName(e) + "]");
        ++fails_;
        cv_.notify_all();
    }
    bool wait(int32_t ms) {
        std::unique_lock<std::mutex> lk(m_);
        return cv_.wait_for(lk, std::chrono::milliseconds(ms),
                            [this] { return oks_ + fails_ > 0; });
    }
    size_t oks() {
        std::lock_guard<std::mutex> lk(m_);
        return oks_;
    }
    SendResult result() {
        std::lock_guard<std::mutex> lk(m_);
        return result_;
    }
    std::vector<std::string> errors() {
        std::lock_guard<std::mutex> lk(m_);
        return errors_;
    }

private:
    std::mutex m_;
    std::condition_variable cv_;
    size_t oks_ = 0;
    size_t fails_ = 0;
    SendResult result_;
    std::vector<std::string> errors_;
};

// 环境：一个名字服务地址 + 一组按 stamp 隔离的 topic，跑完自己删干净。
struct Env {
    std::string nsAddr;
    std::string stamp;
    std::vector<std::string> topics;
    DefaultMQAdminExt admin{"HDRADMIN"};
    std::string brokerAddr;
    std::string brokerName;

    std::string topic(const std::string& kind) {
        std::string t = "Hdr" + kind + "_" + stamp;
        topics.push_back(t);
        return t;
    }

    void start() {
        admin.setNamesrvAddr(nsAddr);
        admin.setTimeoutMillis(10000);
        admin.start();
    }

    void cleanup() {
        for (const std::string& t : topics) {
            try {
                admin.deleteTopic(t);
            } catch (const std::exception& e) {
                std::printf("  [diag] deleteTopic(%s) failed: %s\n", t.c_str(), e.what());
            }
        }
        admin.shutdown();
    }
};

// 读回 broker 上该 topic 的 (read, write) 队列数；还不存在时返回 false。
bool queueNums(Env& env, const std::string& topic, int32_t& read, int32_t& write) {
    try {
        const TopicConfig cfg = env.admin.examineTopicConfig(env.brokerAddr, topic);
        read = cfg.readQueueNums;
        write = cfg.writeQueueNums;
        return true;
    } catch (const std::exception&) {
        return false;
    }
}

// 等 topic 在 broker 上出现（自动建 topic 是发送链路里同步做的，配置落地要一点点时间）
bool waitForTopic(Env& env, const std::string& topic, int32_t& read, int32_t& write) {
    return waitUntil([&] { return queueNums(env, topic, read, write); }, 10000);
}

// 该 topic 全 broker 的落库条数（新 topic 的 minOffset 恒为 0）；读不到返回 -1
int64_t totalMessages(Env& env, const std::string& topic) {
    try {
        const TopicStatsTable stats = env.admin.examineTopicStats(topic);
        int64_t total = 0;
        for (const auto& kv : stats.offsetTable) {
            total += kv.second.maxOffset - kv.second.minOffset;
        }
        return total;
    } catch (const std::exception&) {
        return -1;
    }
}

// 按场景配好并启动一个生产者
class ScopedProducer {
public:
    ScopedProducer(Env& env, const std::string& kind, const std::string& createTopicKey = "",
                   int32_t defaultTopicQueueNums = -1)
        : producer_("PID_send_header_" + env.stamp) {
        producer_.setNamesrvAddr(env.nsAddr);
        producer_.setInstanceName("hdr-" + kind + "-" + env.stamp);
        producer_.setSendMsgTimeout(5000);
        if (!createTopicKey.empty()) producer_.setCreateTopicKey(createTopicKey);
        if (defaultTopicQueueNums >= 0) {
            producer_.setDefaultTopicQueueNums(defaultTopicQueueNums);
        }
        producer_.start();
    }
    ~ScopedProducer() {
        try {
            producer_.shutdown();
        } catch (const std::exception& e) {
            std::printf("  [diag] producer shutdown failed: %s\n", e.what());
        }
    }

    DefaultMQProducer& operator*() { return producer_; }
    SendResult send(const std::string& topic, const std::string& body) {
        return producer_.send(Message(topic, bytesOf(body)), 5000);
    }

private:
    DefaultMQProducer producer_;
};

// ---------------- H0/H1/H2/H3：队列数由 c 和 d 决定 ----------------
void checkQueueNums(Env& env) {
    int32_t tbwRead = 0;
    int32_t tbwWrite = 0;
    if (!queueNums(env, MixAll::DEFAULT_TOPIC, tbwRead, tbwWrite)) {
        check("H0 TBW102 模板可读", false, "examineTopicConfig(TBW102) 读不到");
        return;
    }
    check("H0 TBW102 模板可读", tbwWrite > 0,
          "read=" + num(tbwRead) + " write=" + num(tbwWrite));

    // H1：什么都不设，d 走 MixAll::DEFAULT_TOPIC_QUEUE_NUMS(4)
    const std::string t1 = env.topic("Def");
    {
        ScopedProducer p(env, "h1");
        const SendResult r = p.send(t1, "h1");
        check("H1 默认配置发送成功", r.sendStatus == SendStatus::SEND_OK,
              "msgId=" + r.msgId + " broker=" + r.messageQueue.brokerName);
    }
    int32_t r1 = 0;
    int32_t w1 = 0;
    waitForTopic(env, t1, r1, w1);
    const int32_t expected1 = std::min(MixAll::DEFAULT_TOPIC_QUEUE_NUMS, tbwWrite);
    check("H1 默认 d=4 建的 topic 队列数=min(4, TBW102)", w1 == expected1 && r1 == w1,
          "read=" + num(r1) + " write=" + num(w1) + " (TBW102=" + num(tbwWrite) + ")");

    // H2：d=2 必须把队列数带下去——修之前这里写死 4
    const std::string t2 = env.topic("Nums");
    {
        ScopedProducer p(env, "h2", "", 2);
        p.send(t2, "h2");
    }
    int32_t r2 = 0;
    int32_t w2 = 0;
    waitForTopic(env, t2, r2, w2);
    check("H2 defaultTopicQueueNums=2 建的 topic 只有 2 条队列",
          w2 == std::min(2, tbwWrite) && r2 == w2,
          "read=" + num(r2) + " write=" + num(w2) + "（写死 4 的旧行为会是 " + num(w1) + "）");

    // H3：c 指向自己的模板 topic（必须带 PERM_INHERIT，否则 TopicConfigManager:286 的
    // isInherited 不通过，broker 直接拒绝自动建 topic）
    const std::string src = env.topic("Src");
    try {
        env.admin.createTopicInBroker(env.brokerAddr, src, 3, 3,
                                      PermName::PERM_READ | PermName::PERM_WRITE |
                                          PermName::PERM_INHERIT);
    } catch (const std::exception& e) {
        check("H3 模板 topic 创建", false, e.what());
        return;
    }
    int32_t srcRead = 0;
    int32_t srcWrite = 0;
    waitForTopic(env, src, srcRead, srcWrite);
    check("H3 模板 topic 建好（3 条队列、带 INHERIT）",
          srcWrite == 3 && !(srcRead == tbwRead && srcWrite == tbwWrite),
          "src=(" + num(srcRead) + "," + num(srcWrite) + ") tbw102=(" + num(tbwRead) + "," +
              num(tbwWrite) + ")");

    const std::string t3 = env.topic("Inherit");
    {
        ScopedProducer p(env, "h3", src, 8);
        p.send(t3, "h3");
    }
    int32_t r3 = 0;
    int32_t w3 = 0;
    waitForTopic(env, t3, r3, w3);
    check("H3 createTopicKey=模板 topic 时被继承（min(8,3)=3 而不是 TBW102 的 " +
              num(tbwWrite) + "）",
          w3 == 3 && r3 == 3, "read=" + num(r3) + " write=" + num(w3));
}

// ---------------- H4：五种入口都照常落地 ----------------
void checkSendEntries(Env& env) {
    const std::string t4 = env.topic("Entries");
    std::map<std::string, bool> landed;
    auto latch = std::make_shared<Latch>();
    {
        ScopedProducer p(env, "h4");

        const SendResult rs = p.send(t4, "h4-sync");
        landed["sync"] = rs.sendStatus == SendStatus::SEND_OK;

        // 定点发送：显式给 mq，落在另一条调用链上（broker 名由调用方给）
        const MessageQueue mq = rs.messageQueue;
        const SendResult rb = (*p).send(Message(t4, bytesOf("h4-pinned")), mq, 5000);
        landed["pinned"] =
            rb.sendStatus == SendStatus::SEND_OK && rb.messageQueue.brokerName == mq.brokerName;

        try {
            (*p).sendOneway(Message(t4, bytesOf("h4-oneway")));
            landed["oneway"] = true;  // 单向没有应答，只能靠 H4 的总数兜底
        } catch (const std::exception& e) {
            landed["oneway"] = false;
            std::printf("  [diag] oneway threw: %s\n", e.what());
        }

        std::vector<Message> batch;
        for (int32_t i = 0; i < 3; ++i) {
            batch.push_back(Message(t4, bytesOf("h4-batch-" + std::to_string(i))));
        }
        landed["batch"] = (*p).sendBatch(batch, 5000).sendStatus == SendStatus::SEND_OK;

        (*p).sendAsync(Message(t4, bytesOf("h4-async")), latch, 5000);
        landed["async"] = latch->wait(10000) && latch->oks() == 1 &&
                          latch->result().sendStatus == SendStatus::SEND_OK;
    }
    for (const char* entry : {"sync", "pinned", "oneway", "batch", "async"}) {
        auto it = landed.find(entry);
        const bool ok = it != landed.end() && it->second;
        check(std::string("H4 ") + entry + " 入口发送成功", ok,
              ok ? "" : "该入口没有拿到 SEND_OK");
    }
    // 1 同步 + 1 定点 + 1 单向 + 3 批量 + 1 异步 = 7 条
    const bool got = waitUntil([&] { return totalMessages(env, t4) >= 7; }, 20000);
    const int64_t total = totalMessages(env, t4);
    check("H4 七条消息逐条落库（批量按子消息计）", got && total == 7,
          "total=" + num64(total));
}

// ---------------- H5：落点 broker 名与路由一致 ----------------
void checkBrokerName(Env& env) {
    const std::string t5 = env.topic("BrokerName");
    SendResult r5;
    {
        ScopedProducer p(env, "h5");
        r5 = p.send(t5, "h5");
    }
    check("H5 落点 broker 名与路由一致（n 就是它）",
          r5.sendStatus == SendStatus::SEND_OK && r5.messageQueue.brokerName == env.brokerName,
          "落点=" + r5.messageQueue.brokerName + " 路由=" + env.brokerName);
}

}  // namespace

int main(int argc, char** argv) {
    Env env;
    env.nsAddr = argc > 1 ? argv[1] : "127.0.0.1:9876";
    env.stamp = std::to_string(
        std::chrono::duration_cast<std::chrono::seconds>(
            std::chrono::system_clock::now().time_since_epoch())
            .count());

    try {
        env.start();
    } catch (const std::exception& e) {
        std::printf("admin start failed: %s\n", e.what());
        return 1;
    }

    // 集群探活：拿 broker 地址，并反查它的 brokerName（H5 的判据）
    ClusterInfo cluster;
    for (int32_t i = 0; i < 40; ++i) {
        try {
            cluster = env.admin.fetchBrokerClusterInfo();
            if (!cluster.brokerAddrTable.empty()) break;
        } catch (const std::exception& e) {
            if (i == 39) std::printf("  [diag] cluster probe failed: %s\n", e.what());
        }
        if (!cluster.brokerAddrTable.empty()) break;
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }
    if (cluster.brokerAddrTable.empty()) {
        check("集群探活", false, "nameServer 无 broker 注册");
        env.cleanup();
        return 1;
    }
    const auto it = cluster.brokerAddrTable.begin();
    env.brokerName = it->first;
    env.brokerAddr = it->second.selectBrokerAddr();
    check("集群探活", true, "broker=" + env.brokerAddr + " name=" + env.brokerName);

    std::printf("\n-- H0~H3: c/d 决定自动建 topic 的队列数 --\n");
    checkQueueNums(env);
    std::printf("\n-- H4: 五种发送入口逐条落地 --\n");
    checkSendEntries(env);
    std::printf("\n-- H5: n 就是路由选中的那台 broker --\n");
    checkBrokerName(env);

    env.cleanup();
    std::printf("\n== 结果: %d/%d 通过 ==\n", gPass, gPass + gFail);
    return gFail == 0 ? 0 : 1;
}

File: dotnet/examples/RocketMQ.Examples/LiveSendHeader.cs
// 发送头三个字段（defaultTopic / defaultTopicQueueNums / brokerName）真机联调
//（对应 python/verify_send_header_live.py、cpp/examples/live_send_header.cpp、
//  rust/examples/live_send_header.rs 的 H0–H5）。
//
// 离线单测（tests/SendRetryTests.cs）锁的是**上线形状**：V2 头的单字母键 c/d/n 有没有
// 真的写进 extFields、五种发送入口是不是带同一份值。但「字段上了线」和「broker 真拿它做
// 了决定」是两件事，后者只有真集群能证：
//   H1  默认值：不带任何配置发到新 topic，broker 按 min(d=4, TBW102.writeQueueNums) 建出
//       队列（TopicConfigManager.java:289），与 Java 客户端默认行为一致。
//   H2  DefaultTopicQueueNums=2 真的生效：建出来的 topic 只有 2 条队列。修之前这里写死 4，
//       这条必然变 4 —— 那是那个假 setter 唯一可观测的后果。
//   H3  CreateTopicKey=src 真的生效：先建一个带 PERM_INHERIT、3 条队列的模板 topic，再以
//       它为 c 发送 → 新 topic 继承模板的 3 条队列，而不是 TBW102 的 8 条。
//   H4  补上这三个字段之后，五种入口（同步 / 定点 / 单向 / 批量 320 / 异步）在真 broker
//       上仍逐条落地，条数一条不差。
//   H5  n（brokerName）：落点就是路由选中的那台 broker 名。⚠ 经典 broker 的发送链路里
//       没有 requestHeader.getBrokerName() 的读者（5.5.1 源码 grep 过），所以 n 在线上的
//       存在只能由离线抓帧证明，这里不假装能观测到它。
//
// 用法：rmq send-header [namesrv]（需本地 5.5.1 集群，autoCreateTopicEnable=true）
using System.Text;

using RocketMQ.Client;
using RocketMQ.Common;
using RocketMQ.Remoting.Protocol;

namespace RocketMQ.Examples;

public static class LiveSendHeader
{
    private static string _namesrv = "127.0.0.1:9876";
    private static string _stamp = string.Empty;
    private static DefaultMQAdminExt _admin = null!;
    private static string _brokerAddr = string.Empty;
    private static string _brokerName = string.Empty;

    private static int _pass;
    private static int _fail;

    private static void Check(string name, bool ok, string detail = "")
    {
        if (ok)
        {
            Interlocked.Increment(ref _pass);
            Console.WriteLine("  [PASS] " + name + (detail.Length > 0 ? "  " + detail : ""));
        }
        else
        {
            Interlocked.Increment(ref _fail);
            Console.WriteLine("  [FAIL] " + name + (detail.Length > 0 ? "  " + detail : ""));
        }
    }

    private static byte[] Bytes(string s) => Encoding.UTF8.GetBytes(s);

    private static void Sleep(int ms) => Thread.Sleep(ms);

    private static bool WaitUntil(Func<bool> pred, int seconds)
    {
        DateTime deadline = DateTime.UtcNow.AddSeconds(seconds);
        for (;;)
        {
            if (pred()) return true;
            if (DateTime.UtcNow >= deadline) return false;
            Sleep(100);
        }
    }

    /// <summary>按 stamp 隔离的一组 topic；跑完 Cleanup 扫干净。</summary>
    private static readonly List<string> Topics = new();

    private static string Topic(string kind)
    {
        string t = "Hdr" + kind + "_" + _stamp;
        Topics.Add(t);
        return t;
    }

    /// <summary>一次异步发送的终态。</summary>
    private sealed class Latch : ISendCallback
    {
        private readonly object _lk = new();
        private readonly List<SendResult> _oks = new();
        private readonly List<Exception> _errors = new();

        public void OnSuccess(SendResult sendResult)
        {
            lock (_lk) _oks.Add(sendResult);
        }

        public void OnException(Exception error)
        {
            lock (_lk) _errors.Add(error);
        }

        public bool WaitDone(int ms) => WaitUntil(() =>
        {
            lock (_lk) return _oks.Count + _errors.Count > 0;
        }, ms / 1000);

        public bool FirstOk()
        {
            lock (_lk) return _oks.Count == 1 && _errors.Count == 0
                && _oks[0].SendStatus == SendStatus.SendOk;
        }

        public string Errors()
        {
            lock (_lk) return string.Join(";", _errors.Select(e => e.Message));
        }
    }

    /// <summary>
    /// 配好并启动一个生产者。instanceName 显式给定，clientId 才可预测；c/d 两个旋钮
    /// 传 null 表示保持默认（H1 就是拿默认值当基线的）。
    /// </summary>
    private sealed class ScopedProducer : IDisposable
    {
        private readonly DefaultMQProducer _producer;

        public ScopedProducer(string kind, string? createTopicKey = null,
            int? defaultTopicQueueNums = null)
        {
            _producer = new DefaultMQProducer("PID_send_header_" + _stamp)
            {
                NamesrvAddr = _namesrv,
                InstanceName = "hdr-" + kind + "-" + _stamp,
                SendMsgTimeout = 5000,
            };
            if (createTopicKey is not null) _producer.CreateTopicKey = createTopicKey;
            if (defaultTopicQueueNums is not null)
            {
                _producer.DefaultTopicQueueNums = defaultTopicQueueNums.Value;
            }
            _producer.Start();
        }

        public DefaultMQProducer Producer => _producer;

        public SendResult Send(string topic, string body) =>
            _producer.Send(new Message(topic, Bytes(body)), 5000);

        public void Dispose()
        {
            try
            {
                _producer.Shutdown();
            }
            catch (Exception e)
            {
                Console.WriteLine("  [diag] producer shutdown failed: " + e.Message);
            }
        }
    }

    /// <summary>读回 broker 上该 topic 的 (read, write) 队列数；还不存在返回 null。</summary>
    private static (int Read, int Write)? QueueNums(string topic)
    {
        try
        {
            TopicConfig cfg = _admin.ExamineTopicConfig(_brokerAddr, topic);
            return (cfg.ReadQueueNums, cfg.WriteQueueNums);
        }
        catch (Exception)
        {
            return null;
        }
    }

    /// <summary>等 topic 在 broker 上出现（自动建 topic 在发送链路里同步做，落地要一点时间）。</summary>
    private static (int Read, int Write) WaitForTopic(string topic)
    {
        WaitUntil(() => QueueNums(topic) is not null, 10);
        return QueueNums(topic) ?? (-1, -1);
    }

    /// <summary>该 topic 全 broker 的落库条数（新 topic 的 minOffset 恒为 0）；读不到返回 -1。</summary>
    private static long TotalMessages(string topic)
    {
        try
        {
            TopicStatsTable stats = _admin.ExamineTopicStats(topic);
            long total = 0;
            foreach (KeyValuePair<MessageQueue, TopicOffset> kv in stats.OffsetTable)
            {
                total += kv.Value.MaxOffset - kv.Value.MinOffset;
            }
            return total;
        }
        catch (Exception)
        {
            return -1;
        }
    }

    // ------------------------------------------------- H0/H1/H2/H3 队列数
    private static void H0ToH3QueueNums()
    {
        // H0：TBW102 是这个 broker 上真正的「模板 topic」，H2/H3 的判据都依赖它的队列数
        (int Read, int Write)? tbw = QueueNums(MixAll.DefaultTopic);
        if (tbw is null || tbw.Value.Write <= 0)
        {
            Check("H0 TBW102 模板可读", false, "examineTopicConfig(TBW102) 读不到");
            return;
        }
        Check("H0 TBW102 模板可读", true, "read=" + tbw.Value.Read + " write=" + tbw.Value.Write);

        // H1：什么都不设，d 走 MixAll.DefaultTopicQueueNums(4)
        string t1 = Topic("Def");
        using (ScopedProducer p1 = new("h1"))
        {
            SendResult r1 = p1.Send(t1, "h1");
            Check("H1 默认配置发送成功", r1.SendStatus == SendStatus.SendOk,
                "msgId=" + r1.MsgId + " broker=" + r1.MessageQueue.BrokerName);
        }
        (int Read, int Write) n1 = WaitForTopic(t1);
        int want1 = Math.Min(MixAll.DefaultTopicQueueNums, tbw.Value.Write);
        Check("H1 默认 d=4 建的 topic 队列数=min(4, TBW102)",
            n1.Write == want1 && n1.Read == n1.Write,
            "read=" + n1.Read + " write=" + n1.Write + " (TBW102=" + tbw.Value.Write + ")");

        // H2：d=2 必须把队列数带下去——修之前这里写死 4
        string t2 = Topic("Nums");
        using (ScopedProducer p2 = new("h2", null, 2))
        {
            p2.Send(t2, "h2");
        }
        (int Read, int Write) n2 = WaitForTopic(t2);
        Check("H2 DefaultTopicQueueNums=2 建的 topic 只有 2 条队列",
            n2.Write == Math.Min(2, tbw.Value.Write) && n2.Read == n2.Write,
            "read=" + n2.Read + " write=" + n2.Write + "（写死 4 的旧行为会是 " + n1.Write + "）");

        // H3：c 指向自己的模板 topic。模板必须带 PERM_INHERIT，否则 TopicConfigManager:286
        // 的 isInherited 不通过，broker 直接拒绝自动建 topic
        string src = Topic("Src");
        try
        {
            _admin.CreateTopicInBroker(_brokerAddr, src, 3, 3,
                PermName.PermRead | PermName.PermWrite | PermName.PermInherit);
        }
        catch (Exception e)
        {
            Check("H3 模板 topic 创建", false, e.Message);
            return;
        }
        (int Read, int Write) nsrc = WaitForTopic(src);
        Check("H3 模板 topic 建好（3 条队列、带 INHERIT）",
            nsrc.Write == 3 && !(nsrc.Read == tbw.Value.Read && nsrc.Write == tbw.Value.Write),
            "src=(" + nsrc.Read + "," + nsrc.Write + ") tbw102=(" + tbw.Value.Read + "," +
            tbw.Value.Write + ")");

        string t3 = Topic("Inherit");
        using (ScopedProducer p3 = new("h3", src, 8))
        {
            p3.Send(t3, "h3");
        }
        (int Read, int Write) n3 = WaitForTopic(t3);
        Check("H3 createTopicKey=模板 topic 时被继承（min(8,3)=3 而不是 TBW102 的 " +
            tbw.Value.Write + "）", n3.Write == 3 && n3.Read == 3,
            "read=" + n3.Read + " write=" + n3.Write);
    }

    // ------------------------------------------------------------- H4 五种入口
    private static void H4SendEntries()
    {
        string t4 = Topic("Entries");
        Dictionary<string, string> landed = new();
        Latch latch = new();
        using (ScopedProducer p = new("h4"))
        {
            SendResult rs = p.Send(t4, "h4-sync");
            landed["sync"] = rs.SendStatus == SendStatus.SendOk ? "" : "status=" + rs.SendStatus;

            // 定点发送：显式给 mq，落在另一条调用链上（broker 名由调用方给）
            MessageQueue mq = rs.MessageQueue;
            SendResult rb = p.Producer.Send(new Message(t4, Bytes("h4-pinned")), mq, 5000);
            landed["pinned"] = rb.SendStatus == SendStatus.SendOk
                && rb.MessageQueue.BrokerName == mq.BrokerName
                ? "" : "status=" + rb.SendStatus + " broker=" + rb.MessageQueue.BrokerName;

            try
            {
                p.Producer.SendOneway(new Message(t4, Bytes("h4-oneway")));
                landed["oneway"] = "";  // 单向没有应答，只能靠后面的总数兜底
            }
            catch (Exception e)
            {
                landed["oneway"] = e.Message;
            }

            List<Message> batch = new();
            for (int i = 0; i < 3; i++) batch.Add(new Message(t4, Bytes("h4-batch-" + i)));
            SendResult rr = p.Producer.SendBatch(batch, 5000);
            landed["batch"] = rr.SendStatus == SendStatus.SendOk ? "" : "status=" + rr.SendStatus;

            p.Producer.SendAsync(new Message(t4, Bytes("h4-async")), latch, 5000);
            landed["async"] = latch.WaitDone(10000) && latch.FirstOk()
                ? "" : "回调没拿到唯一的 SEND_OK：" + latch.Errors();
        }

        foreach (string entry in new[] { "sync", "pinned", "oneway", "batch", "async" })
        {
            string why = landed.TryGetValue(entry, out string? v) ? v : "该入口没有记录";
            Check("H4 " + entry + " 入口发送成功", why.Length == 0, why);
        }
        // 1 同步 + 1 定点 + 1 单向 + 3 批量 + 1 异步 = 7 条
        bool got = WaitUntil(() => TotalMessages(t4) >= 7, 20);
        Check("H4 七条消息逐条落库（批量按子消息计）", got && TotalMessages(t4) == 7,
            "total=" + TotalMessages(t4));
    }

    // ------------------------------------------------------------- H5 brokerName
    private static void H5BrokerNameMatchesRoute()
    {
        string t5 = Topic("BrokerName");
        SendResult r5;
        using (ScopedProducer p = new("h5"))
        {
            r5 = p.Send(t5, "h5");
        }
        Check("H5 落点 broker 名与路由一致（n 就是它）",
            r5.SendStatus == SendStatus.SendOk && r5.MessageQueue.BrokerName == _brokerName,
            "落点=" + r5.MessageQueue.BrokerName + " 路由=" + _brokerName);
    }

    private static void Cleanup()
    {
        foreach (string t in Topics)
        {
            try
            {
                _admin.DeleteTopic(t);
            }
            catch (Exception e)
            {
                Console.WriteLine("  [diag] DeleteTopic(" + t + ") failed: " + e.Message);
            }
        }
    }

    public static int Run(string[] args)
    {
        _namesrv = args.Length > 0 ? args[0] : "127.0.0.1:9876";
        _stamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds().ToString();
        Console.WriteLine("namesrv=" + _namesrv + " stamp=" + _stamp);

        _admin = new DefaultMQAdminExt("HDRDOTNETADMIN");
        _admin.SetNamesrvAddr(_namesrv);
        _admin.SetTimeoutMillis(10000);
        _admin.Start();

        try
        {
            // 集群探活：拿 broker 地址，并反查它的 brokerName（H5 的判据）
            ClusterInfo? cluster = null;
            WaitUntil(() =>
            {
                try
                {
                    cluster = _admin.FetchBrokerClusterInfo();
                    return cluster.BrokerAddrTable.Count > 0;
                }
                catch (Exception)
                {
                    return false;
                }
            }, 30);
            if (cluster is null || cluster.BrokerAddrTable.Count == 0)
            {
                Check("集群探活", false, "nameServer 无 broker 注册");
            }
            else
            {
                foreach (KeyValuePair<string, BrokerData> kv in cluster.BrokerAddrTable)
                {
                    _brokerAddr = kv.Value.SelectBrokerAddr();
                    _brokerName = kv.Key;
                    break;
                }
                Check("集群探活", _brokerAddr.Length > 0,
                    "broker=" + _brokerAddr + " name=" + _brokerName);

                H0ToH3QueueNums();
                H4SendEntries();
                H5BrokerNameMatchesRoute();
            }
        }
        catch (Exception e)
        {
            Check("联调异常", false, e.Message);
        }

        Cleanup();
        _admin.Shutdown();
        Console.WriteLine();
        Console.WriteLine("PASS=" + Volatile.Read(ref _pass) + " FAIL=" + Volatile.Read(ref _fail));
        return Volatile.Read(ref _fail) == 0 ? 0 : 1;
    }
}
